### PR TITLE
Improve DinoV2 Fine-Tuning for Classification Performance  

### LINKED ISSUE #4288: Improve DinoV2 performance for classification tasks
We have an existing implementation of DinoV2 (https://github.com/openvinotoolkit/training_extensions/blob/develop/src/otx/algo/classification/backbones/vision_transformer.py) in our repository, but it currently underperforms compared to vanilla CNNs in classification tasks, both in terms of accuracy and efficiency. This issue aims to investigate and improve the fine-tuning process of DinoV2, specifically focusing on fine-tuning only the linear classification head. However, full fine-tuning of the model can also be considered as a baseline for comparison.

Tasks:
1. Analyze Current Performance:

  -  Evaluate the accuracy and efficiency of the current DinoV2 implementation.
  -  Compare results with a "standard" CNN baseline (consider our implemented models in OTX like MobileNetV3, EfficientNetb0, EfficientNetV2-s, etc.).
2. Optimize Backbone and Linear Head Fine-Tuning: 

  - Identify potential improvements in the fine-tuning setup (e.g., learning rate, optimizer, data augmentation, loss function, input size and embedding dimension, etc.).
  - Implement changes while keeping them compatible with the current repository design.

 3. Consider Full Fine-Tuning (Baseline):

-  Experiment with full fine-tuning of DinoV2 as a baseline to compare performance gains.

 4. Benchmark and Report Results:
  -  Measure improvements in accuracy and efficiency. At least 4-5 diverse datasets should be considered with different number of images.
  -  Provide recipe (config) for e2e working solution (from torch fine-tuning to OpenVINO IR model)

### PR DESCRIPTION
**Pull request by faizan102418**
### Summary

Resolves #4288 .

This PR improves the fine-tuning process of DinoV2 for classification tasks in OpenVINO Training Extensions. The current DinoV2 implementation underperforms compared to CNNs, so we introduce the following optimizations:

- Enhancements in Data Preprocessing & Augmentations
- Increased input size from 224x224 to 256x256 for better feature extraction.
- Enabled stronger data augmentations (MixUp, CutMix, RandomAffine, GaussianNoise, etc.) to improve generalization.
- Enhanced normalization to stabilize training
- Optimization in Training Configuration
- Improved learning rate scheduling for better convergence.
- Fine-tuned batch size and embedding dimensions for optimal memory utilization.
- Expected Outcome:
    -Improved validation accuracy and test accuracy for DinoV2.
    -Better generalization across diverse datasets.
    -Competitive performance against EfficientNets and MobileNet-V3.

### License

- [x ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/recipe/_base_/data/classification.yaml [hunk 1]
```diff
@@ -18,21 +19,31 @@ train_subset:
...
         is_numpy_to_tvtensor: true
     - class_path: torchvision.transforms.v2.RandomVerticalFlip
-      enable: false
+      enable: true  # Enabled vertical flipping
     - class_path: torchvision.transforms.v2.GaussianBlur
-      enable: false
+      enable: true
+      init_args:
+        kernel_size: 3  # Reduced blur intensity
+    - class_path: torchvision.transforms.v2.GaussianNoise
```
Comment (kprokofi) [inline src/otx/recipe/_base_/data/classification.yaml]: It seems like a lot of augmentations for me. I think there should be an ablation study if we really need all of them.

File: src/otx/recipe/_base_/data/classification.yaml [hunk 2]
```diff
@@ -55,7 +65,7 @@ val_subset:
   transforms:
     - class_path: otx.core.data.transform_libs.torchvision.Resize
       init_args:
-        scale: $(input_size)
+        scale: [256, 256]  # Updated input size
```
Comment (kprokofi) [inline src/otx/recipe/_base_/data/classification.yaml]: please, use $(input_size). It will be automatically set to the right input size based on the provided input size to the model. You need to change this only on the 3-4 lines

File: src/otx/recipe/classification/multi_class_cls/dino_v2.yaml [hunk 3]
```diff
@@ -7,20 +7,18 @@ model:
...
-        lr: 0.0001
-        weight_decay: 0.05
+        lr: 0.0008  # Increased learning rate for better training
+        weight_decay: 0.02  # Reduced weight decay
 
     scheduler:
       class_path: otx.core.schedulers.LinearWarmupSchedulerCallable
       init_args:
-        num_warmup_steps: 0
+        num_warmup_steps: 5  # Small warm-up for stability
```
Comment (kprokofi) [inline src/otx/recipe/classification/multi_class_cls/dino_v2.yaml]: please note, it is not epochs, it is iterations. 5 iterations seems to be unreasonable small value for the warm-up. However, since we are working with small datasets, we could experiment with this value up to 100-300 iterations.

File: src/otx/recipe/classification/multi_class_cls/dino_v2.yaml [hunk 4]
```diff
@@ -7,20 +7,18 @@ model:
...
         main_scheduler_callable:
-          class_path: lightning.pytorch.cli.ReduceLROnPlateau
+          class_path: torch.optim.lr_scheduler.CosineAnnealingLR
           init_args:
-            mode: max
-            factor: 0.5
-            patience: 3
-            monitor: val/accuracy
+            T_max: 120
```
Comment (kprokofi) [inline src/otx/recipe/classification/multi_class_cls/dino_v2.yaml]: Please note, it is iterations, not epochs. 120 iterations for scheduler is unreasonable value.

**Review by kprokofi — COMMENTED** [review]
Please, provide ablation study and comparison benchmark results for your changes.

**Comment by tech-wizar** [comment]
take

File: src/otx/recipe/classification/multi_class_cls/dino_v2.yaml (hunk 4 above)
Comment (faizan102418) [reply src/otx/recipe/classification/multi_class_cls/dino_v2.yaml]: Let me Improve It, and thank you for your guidance and precious support @sovrasov and share my progress with you.

**Comment by RoushanRaj-cmd** [comment]
. take

File: src/otx/algo/classification/backbones/vision_transformer.py [hunk 5]
```diff
@@ -124,9 +124,10 @@ class VisionTransformer(BaseModule):
                 "patch_size": 14,
                 "embed_dim": 384,
                 "depth": 12,
-                "num_heads": 6,
```
Comment (sovrasov) [inline src/otx/algo/classification/backbones/vision_transformer.py:127]: Would that change also affect DeiT OTX model?

File: src/otx/recipe/_base_/data/classification.yaml [hunk 6]
```diff
@@ -1,4 +1,4 @@
-task: MULTI_CLASS_CLS
+task: MULTI_CLASS_CLS
```
Comment (sovrasov) [inline src/otx/recipe/_base_/data/classification.yaml:1]: This file is common for all classification models, while we're aiming only DinoV2 in this PR. I'd propose to make all the required overrides in the Dino's recipe, or make a Dino-specific data config alongside with the `classification.yaml`.

**Comment by sovrasov** [comment]
@faizan102418 please run `tox -vv -e pre-commit` to fix minor linter issues.

**Comment by dawoodshahzad07** [comment]
.take

File: src/otx/algo/classification/backbones/vision_transformer.py (hunk 5 above)
Comment (faizan102418) [reply src/otx/algo/classification/backbones/vision_transformer.py:127]: I am working on it,  and will update you about this

File: src/otx/recipe/classification/multi_class_cls/dino_v2.yaml (hunk 3 above)
Comment (faizan102418) [reply src/otx/recipe/classification/multi_class_cls/dino_v2.yaml]: Got it, Thanks

**Comment by raahulsaxena** [comment]
I wanted to request the assignment of this issue to me!

**Comment by faizan102418** [comment]
Hi @sovrasov ,
I am working on improving the DinoV2 fine-tuning setup and testing different adjustments. I'll share my progress soon. Let me know if you have any suggestions!
BR,
Muhammad Faizan Sajid

**Comment by PraroopChanda** [comment]
.take

**Comment by PraroopChanda** [comment]
Hi @sovrasov , 
Would really like to request the assignment of this issue.

**Comment by isthatdeepti** [comment]
.take

**Comment by EdOgbalor** [comment]
.take

**Comment by kprokofi** [comment]
Close due to inactivity
